syntax = "proto3";

package deepx;

message Discrete {
    int32 n = 1;
}

message Box {
    repeated int32 shape = 1;
    repeated float low = 2;
    repeated float high = 3;
}

message Space {
    Discrete discrete = 1;
    Box box = 2;
}

message State {
    repeated float obs = 1;
}

message Step {
    State state = 1;
    float reward = 2;
    bool done = 3;
}

message EnvConfig {
    string env_id = 1;
    int32 num_envs = 2;
    Space action_space = 3;
    Space observation_space = 4;
}

message CreateRequest {
    string id = 1;
    int32 num_envs = 2;
}

message CreateResponse {
    EnvConfig config = 1;
}

message ResetRequest {
    string env_id = 1;
}

message ResetResponse {
    repeated State states = 1;
}

message StepRequest {
    string env_id = 1;
    repeated int32 action = 2;
}

message StepResponse {
    repeated Step steps = 1;
}

message CloseRequest {
    string env_id = 1;
}

message ListRequest {}

message ListResponse {
    repeated string env_ids = 1;
}

message CloseResponse {}

service Gym {
    rpc Create(CreateRequest) returns (CreateResponse);
    rpc Reset(ResetRequest) returns (ResetResponse);
    rpc Step(StepRequest) returns (StepResponse);
    rpc Close(CloseRequest) returns (CloseResponse);
    rpc List(ListRequest) returns (ListResponse);
}